### ISSUE TITLE
Feature Request: Web Speech API support

### ISSUE THREAD

**Issue by 3rd-Eden** [DestinyItemManager/DIM#2495]
Yes, you've read it correctly. Web. Speech.

The original inspiration comes from the [Alexa ghost](https://www.amazon.co.uk/Bungie-Destiny-2-Ghost/dp/B077HD31C4) that allows Alexa commands to interact with the game. The concept here is great and could be applied to DIM.

As DIM currently runs within a browser the build-in Speech API could be leveraged to provide similar functionality. When running a couple of example commands through a [Speech demo](https://www.google.com/intl/en/chrome/demos/speech.html) you will notice that the transformation is pretty damn close to what is said.

Ofcourse this technology is not foolproof, nor would it work for every single command you barf at it but initial implementations could start with some basic commands such as:

- `DIM, Move sins of the past`
- `DIM, Make room in inventory`
- `DIM, Start farm mode`
- `DIM, Equip loadout 1`

There are great natural speech libraries available that implement soundex, metaphone, double metaphone and other algorithms to aid with mapping speech to commands. The most obvious library would be [natural](https://github.com/NaturalNode/natural). 

I do fully understand that this will not be small undertaking, but in my opinion it would be worth the exploring and could be another unique selling point of DIM. Use cases would be running DIM in the background on PC, with the microphone open. Listening to commands based on the DIM keyword so no switching is needed.

**Comment by bhollis** [comment]
Our effort on this is at https://github.com/DestinyItemManager/iot, though the initial steam died out especially when the official Alexa skill showed up.

**Comment by SunburnedGoose** [comment]
Now this is an interesting idea that we can bake directly into DIM.

**Comment by 3rd-Eden** [comment]
It's also a lot easier to come with a MVP instead of creating and API for IoT. The only thing is it will be only be available for browsers that support the Speech API. If you're interested in adding functionality in to DIM, I wouldn't mind creating a standalone small PoC that transforms some Speech examples in a JSON/API structure.

**Comment by SunburnedGoose** [comment]
Let's do it.

**Comment by nki2bpz** [comment]
"Our effort on this is at https://github.com/DestinyItemManager/iot, though the initial steam died out especially when the official Alexa skill showed up." This part is very sad. Anyone who lost steam from the official skill being released has never used the official skill. It is truly abysmal. It sounds good cuz the ghost voice but has almost no functionality at all. You can save 3 loadouts of 3 weapons each and it can not move a requested weapon by name. I still dream of the day I can say "echo, ask dim to equip my crucible loadout" and get 9 weapons transferred.

**Comment by renzdev** [comment]
For what it's worth, I've been looking at a proof of concept for this as a potential project sometime over the next 4-8 weeks.  I'm looking for an excuse to practice with golang and I believe the chrome speech API has a go client library, but I'm not sure how successfully this would interop with the base DIM nodejs implementation? Also, I didn't fully understand the response from @3rd-Eden above, so was waiting for the past week or two to see if there would be any progress from someone else as I didn't want to duplicate effort.

**Comment by SunburnedGoose** [comment]
@3rd-Eden seems interested in building this into DIM's client side application, which is JS that is executed within the browser afaik.  That sounds exciting as it can be local to the users experience, and we can avoid having an external application to manage the IoT implementation (Alexa, Google Home, etc).  There are issues with keeping DIM and changes made with the IoT device in sync for the user w/ refreshing.

**Comment by 3rd-Eden** [comment]
@SunburnedGoose Is correct, I've started working on demo that explores this area. As part of this exploration I've started writing a JSGF document which can be leveraged as independent textual representation of grammars for use in speech recognition so ideally it could be shared between web and other IoT devices. So if anything, completing a web integration first would make it easier to also continue efforts for the IoT project.
